feat(protocol-engine): Add a command queue worker class

amitlissack opened the pull request:
# Overview

This PR adds a `CommandQueueWorker` class to help play and pause a queue of commands that's already in a `ProtocolEngine`'s queue.  

When `play` is called, a background task is created that iterates through pending commands in a `ProtocolEngine`'s state store. Each pending command is executed.

`pause` terminates the task.

closes #7646.

# Changelog

- added `CommandQueueWorker` class.
- added un-implemented `get_next_pending` to command state store.

# Review requests


# Risk assessment

None. Not production code.

Commit:
fixtures and command queue worker tests.

sanni-t committed:
added JsonFileRunner load() method and tests

sanni-t committed:
fix linter errors

sanni-t committed:
add decoy.mypy plugin

sanni-t committed:
one linter error

sanni-t committed:
ugh D403

sanni-t committed:
remove translation test, remove cast, add note about command_ids

sanni-t committed:
fix docstring

Commit:
CommandQueueWorker.

celsasser commented on api/src/opentrons/file_runner/json_file_runner.py:
```diff
@@ -1,18 +1,49 @@
 class JsonFileRunner(AbstractFileRunner):
     """JSON protocol file runner."""
 
+    def __init__(
```
Is this picked up from @sanni-t 's recent PR - https://github.com/Opentrons/opentrons/pull/7797/files#diff-6b95c553d2325826d1e5df2572635789e511e6cf82c9a9985d6e6b1e311aeab4R10?  (maybe should read the PR...). If so then would be very nice to get her's in if possible to reduce the noise.

amitlissack replied:
Oops. I should re-target. I always make the same mistake.

SyntaxColoring committed:
Flesh out docstrings (characterizing existing behavior).

SyntaxColoring commented on api/src/opentrons/file_runner/command_queue_worker.py:
```diff
@@ -1,17 +1,91 @@
+            # that we haven't yet seen.
+            self._task = None
+
+    async def wait_terminated(self) -> None:
+        """Wait until this `CommandQueueWorker` has stopped executing new commands.
```
Would a name like `wait_to_be_idle()` be more representative of what this method does?

SyntaxColoring replied:
Doing this.

SyntaxColoring commented on api/src/opentrons/file_runner/command_queue_worker.py:
```diff
@@ -1,17 +1,91 @@
+        """
+        if self._task is not None:
+            self._terminate = True
+            # fixme(mm, 2021-05-18): I think this is a bug if stop() is called before
+            # wait_terminated(). It discards the background task before it's awaited,
+            # which both leaks the task and throws away any errors raised in the task
+            # that we haven't yet seen.
+            self._task = None
```
Not sure what to do about this, yet.

Does everything work out of we just remove the `self._task = None`?

It might turn out to be moot when we see what we need out of this class when it's used in context, so I'm avoiding thinking about it too hard for now.

SyntaxColoring replied:
I lost track of this through subsequent rewrites, but I *think* no problem like this exists in the current code.

SyntaxColoring committed:
Keep task around until calling `wait_terminated()`.

This is important to propagate errors.

There's a similar potential problem in `stop()`, but that's more subtle, so I'm leaving it alone for now.

SyntaxColoring commented on api/src/opentrons/file_runner/command_queue_worker.py:
```diff
@@ -1,17 +1,91 @@
+    async def _loop(self) -> None:
+        """Loop through pending commands and execute them."""
+        while self._terminate is False:
+            pending = self._engine.state_store.commands.get_next_request()
+            if pending is None:
+                break
+            cmd_id, request = pending
+            await self._engine.execute_command(
+                command_id=cmd_id,
+                request=request
+            )
```
Unclear to me: What happens (or should happen) if a command raises an error? Is that command returned to the front of the queue, or is it consumed?

mcous replied:
Let's model existing behavior: if a command raises an error, the run is aborted. This doesn't need to be handled here; the queue can be configured to return no pending commands if there is a `FailedCommand` in the queue

SyntaxColoring replied:
Edit: I forgot `get_next_request()` is currently totally unimplemented. I left a note in its docstring for this requirement. We'll do this in another PR.

mcous commented on api/src/opentrons/file_runner/command_queue_worker.py:
```diff
@@ -1,17 +1,89 @@
     def pause(self) -> None:
         """Pause queued command execution."""
-        raise NotImplementedError()
+        self.stop()
```
Should `pause` and `stop` have different behaviors? For example:

- `pause`: pause queue processing
- `stop`: marked current command (if any) as `Cancelled`, pause queue processing
    - Perhaps this indicates that the engine needs a `cancel_command` by ID method

SyntaxColoring replied:
As discussed in a call (I think): we'll delete one for now, since it's anticipating a difference that doesn't actually exist today.

Your thoughts about the engine growing a `cancel_command(id)` method sound good to me.

mcous commented on api/src/opentrons/file_runner/command_queue_worker.py:
```diff
@@ -1,17 +1,89 @@
+        """
+        if self._task is None:
+            self._terminate = False
+            self._task = asyncio.create_task(self._loop())
```
I'm fairly partial to a different sort of algorithm for these sorts of loops, where the worker has two methods: one to do the work, and one to schedule the next unit of work, if needed.

- `_execute_command`
    2. `await` the command's execution 
    3. Call `self.schedule_next_command` when the execution is done
- `_schedule_next_command`
    1. Check if the worker should be running
    2. Get next pending command ID
    3. Call `self._task = create_task(self._execute_command(command_id))` if there's a command in the queue
        - Else do `self._task = None` or something 

The reason I like this is that we can separate out execution from scheduling, and also we don't have to write `break` anywhere because there's no while loop. Instead of one big `self._task` the represents the loop, you end up with a self-perpetuating series of `self._task` values, where the current task is always **the execution of the current command**

SyntaxColoring replied:
As we sort of talked about in a call and DMs, I'm not doing this in this PR, because its interaction with how the rest of the class is implemented would cause a bug.

`wait_to_be_idle()` is currently implemented roughly like:

```python
async def wait_to_be_idle(self) -> None:
    if self._task is not None:
        await self._task
        self._task = None
```

The bug is that this would only wait for the *first* task (or some nondeterministic task) to complete, not the *final* task. So, if you `.play()` and then immediately `.wait_to_be_idle()`, `.wait_to_be_idle()` could return after only the *first command* completes, while the rest of the commands are still running.

We could rewrite `wait_to_be_idle()` to fix this. I only got halfway-ish into this before abandoning it, but I think it would require adding an `asyncio.Event` class attribute, and `await`ing that instead of a task to wait for idleness. My hesitations about this are:

* It adds another piece of confusingly quasi-orthogonal state to the class. Although we could mitigate this by having a more structured, dedicated `WorkerState` type instead of a bunch of loose attributes.
* It's difficult (impossible?) to ensure that every task in that self-perpetuating chain is `await`ed. A consequence of not `await`ing a task is that exceptions won't be propagated properly, potentially making errors turn into hangs, or otherwise leaving things in an inconsistent state. (A lot of my fear of un-`await`ed tasks comes from the essay [Notes on structured concurrency, or: Go statement considered harmful](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful).)

So I'm going back to the original approach of having a single task with a loop.

----

I did find a way to rewrite the `while` loop as a `for` loop, though. Maybe that partially addresses your original concern that `while` loops with `break`s don't feel fail-safe?

mcous commented on api/src/opentrons/file_runner/command_queue_worker.py:
```diff
@@ -1,17 +1,89 @@
+        errors.
+
+        Raises:
+            ProtocolEngineError: If there was a problem executing one of the commands
```
This isn't true as written (not that it has to remain the way it currently is!). `engine.execute_command` does not raise right now; it captures errors and puts them in state

SyntaxColoring replied:
Thanks. Removed this note.

Other error-handling discussion in this thread: https://github.com/Opentrons/opentrons/pull/7817#discussion_r634727526

mcous commented on api/src/opentrons/file_runner/command_queue_worker.py:
```diff
@@ -1,17 +1,89 @@
+        If a command is currently executing, this method will return after that command
+        finishes.
+
+        When you're finished with a `CommandQueueWorker`, you must call this method on
```
I don't love this requirement; I think the worker should be able to clean up after itself. See my comment about the task chain for an idea about how that can be accomplished

SyntaxColoring replied:
Discussion of the task chain idea: https://github.com/Opentrons/opentrons/pull/7817#discussion_r636174409

Notes from call:

* The cleanup requirement turns out to be approximately the same whether we have a single task or a self-perpetuating chain of tasks.
* Arguably, there is no cleanup requirement in *either* case; it's not like we need to free a file descriptor, or something.
* However, I eel fairly strongly that if you spawn an asyncio task, you need to make sure it's `await`ed at some point. This makes sure the language's normal mechanisms of exception propagation and resource cleanup work properly. I stole this opinion from [Notes on structured concurrency, or: Go statement considered harmful](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/).

Further thoughts:

Taking for granted that an `await` needs to happen, there's an open interface design question of *where* it should happen, from the `CommandQueueWorker` caller's perspective. Do you call `await wait_to_be_idle()` just once when you're done with the class? Do you call it between each adjacent `pause()` and `play()`? Do we make `pause()` async and bake the `await` into that?

Meanwhile, I softened the documentation from "must" to "should" and left a todo to look into this further.

mcous commented on api/tests/opentrons/file_runner/test_command_queue_worker.py:
```diff
@@ -0,0 +1,121 @@
+    assert protocol_engine.execute_command.call_args_list == expected_call_args_list
+
+
+async def test_pause(
```
This is a pretty funky test (especially `mock_execute_command`), and it says to me that maybe this code could be structured in a way to make this more testable. I think using `decoy` over `MagicMock` could be helpful here, given that it's API is intentionally limited to provide test pain to guide subject structure.

If we ask ourselves "what does pause mean?", I think we only have one real requirement:

- While paused, any attempt to "get the next pending command" should return `None`

I think this test becomes more simple if we spin that functionality off into its own unit, a `CommandQueue` wrapper of the protocol engine. Maybe that `CommandQueue` has three methods:

- `pause`
- `unpause` (or whatever)
- `get_next_command`

Tests for the `CommandQueue` look like:

- Test that it returns the proper command given engine state
- Test that it returns `None` if `pause` was called
- Test that it can be `unpause`'d

Tests for `pause` of the `CommandQueueWorker` then become:

- Test that `worker.pause` and `worker.stop` tell the queue to pause
- Test that `worker.play` tells the queue to unpause
- Ensure that `CommandQueue.get_next_command` is used to feed the next command ID in

That lets you side-step all this lifecycle weirdness and takes `CommandQueueWorker` to a place where it is more strictly a collaborator and not both a thing that delegates _and_ a thing that has logic.

This entire video is great if you haven't already watched it, but it's got [a section specifically about do vs. delegate](https://youtu.be/aTqC5CIy7mg?t=1127).

SyntaxColoring replied:
Notes from call:

* We'll revisit these tests in a future pairing session. Leaving them mostly alone for now.

mcous commented on api/src/opentrons/file_runner/command_queue_worker.py:
```diff
@@ -1,17 +1,89 @@
+    async def _loop(self) -> None:
+        """Loop through pending commands and execute them."""
+        while self._terminate is False:
+            pending = self._engine.state_store.commands.get_next_request()
```
I think it would be better if the command queue worker could deal only in command IDs, i.e. get the ID of the next pending command and tell the engine to execute the pre-existing command with that ID. I think this means re-working or replacing `execute_command`

Because the existing implementation of `execute_command` does too much (it creates _and_ executes the command), this will overwrite the `createdAt` time of the command, which would have been set when it was loaded into the engine.

I think where we currently use `execute_command`, we should replace it with `create_command` from #7797 followed by a new "execute command by ID" method added by this PR

mcous replied:
**Update on this change request: let's punt it.**

In my experimentation this is gnarlier than expected. The PE's design needs to adjust more than expected to account for what had previously only been speculative

mcous reviewed (changes requested):
I've got a fair number of thoughts on this one! Let me know if you'd like to discuss

SyntaxColoring committed:
Merge branch 'edge' into api-command-queue-worker

SyntaxColoring committed:
Merge branch 'edge' into api-command-queue-worker

SyntaxColoring committed:
Import tweak.

SyntaxColoring committed:
Add naming todo.

SyntaxColoring committed:
Rename _terminate to _keep_running.

SyntaxColoring committed:
Rename wait_terminated to wait_to_be_idle.

SyntaxColoring committed:
Document that pause() is equivalent to stop() for now.

SyntaxColoring committed:
Use mutually recursive functions instead of a `break`'d while loop.

SyntaxColoring committed:
Flesh out some docs and todos.

SyntaxColoring committed:
Fix up tests.

SyntaxColoring committed:
Fix CommandQueueWorker initialization in create_file_runner().

SyntaxColoring committed:
Fix race condition where wait_to_be_idle() could await the wrong task.

SyntaxColoring committed:
Soften requirement to call wait_to_be_idle() and leave a todo.

SyntaxColoring committed:
Delete stop(); support only pause().

Preserving pause() over stop() is mostly arbitrary given the my comment about finding more abstraction-appropriate names for everything, anyway. But "pause" connotes resumability to me, which is what calling code currently assumes, I think.

SyntaxColoring commented on api/src/opentrons/file_runner/command_queue_worker.py:
```diff
@@ -1,17 +1,105 @@
+            await self._task
+            self._task = None
+
+    def _next_command(self) -> \
+            Optional[Tuple[str, protocol_engine.commands.CommandRequestType]]:
+        if self._keep_running:
+            # Will be None if the engine has no commands left.
+            return self._engine.state_store.commands.get_next_request()
+        else:
+            return None
```
Given that `CommandRequestType` is part of the public interface of `get_next_request()`, should it be importable as `protocol_engine.CommandRequestType` or something?

Edit: Not dealing with this in this PR, but I'll leave this thread open in case anyone has an opinion.

SyntaxColoring committed:
*very long sigh* trailing whitespace

SyntaxColoring committed:
Clarify docs.

* Clarify that play() and pause() return immediately.
* Improve readability of list.
* Rephrase wait_to_be_idle() docs to avoid accidentally implying that it also automatically acts as a pause.

SyntaxColoring commented on api/src/opentrons/file_runner/command_queue_worker.py:
```diff
@@ -1,17 +1,101 @@
+        executing, it will continue until it's done. Further commands will be left
+        unexecuted in the queue.
+        """
+        if self._task is not None:
```
Do we still need this if we have a single big loop task?

SyntaxColoring replied:
Removed.

SyntaxColoring commented on api/tests/opentrons/file_runner/test_command_queue_worker.py:
```diff
@@ -0,0 +1,136 @@
+    assert protocol_engine.execute_command.call_args_list == expected_call_args_list
+
+
+async def test_stop(
```
Should be `test_pause()`.

mcous reviewed (approved):
LGTM 📦 (discussed a lot in person)

SyntaxColoring committed:
Whitespace fixup.

SyntaxColoring committed:
Rename test_stop() to test_pause().

SyntaxColoring committed:
Remove unnecessary self._task guard.

Clearing self._keep_running *should* be harmless regardless of whether a task is currently running and whether we're storing it in self._task.

SyntaxColoring committed:
Update get_next_request() docstring.

Make it mention a new requirement that we're putting on it: that it must return None if the last command errored.

SyntaxColoring committed:
Consistently use (request, command_id) ordering in tests.

I'm not sure why we were having trouble with this before, but it seems fine now? :|

sanni-t commented on api/tests/opentrons/file_runner/test_command_queue_worker.py:
```diff
@@ -0,0 +1,138 @@
+        subject: CommandQueueWorker,
+        store: AsyncMock
+) -> None:
+    """It not execute any commands."""
```
```suggestion
    """It should not execute any commands."""
```

sanni-t reviewed (approved):
👷‍♂️
Thanks for updating all the comments with details from your discussion!

SyntaxColoring committed:
Add missing word to docstring.

Co-authored-by: Sanniti Pimpley <sanni-t@users.noreply.github.com>
